### PR TITLE
IBP-3202 Add Visualization Tool

### PR DESCRIPTION
**Pull request by abatac**

### CODE REVIEW

File: src/test/java/org/ibp/api/java/impl/middleware/dataset/DatasetServiceImplTest.java [hunk 1]
```diff
@@ -258,6 +258,19 @@ public void testGetObservations() {
 
 	}
 
+	@Test
+	public void testGetObservationUnitRowsAsListMap() {
+		final List<Map<String, Object>> listOfMap = new ArrayList<>();
+
+		Mockito.doReturn(listOfMap).when(this.middlewareDatasetService)
+			.getObservationUnitRowsAsListMap(ArgumentMatchers.anyInt(), ArgumentMatchers.anyInt(), ArgumentMatchers.any());
```
Comment (darla-leafnode) [inline src/test/java/org/ibp/api/java/impl/middleware/dataset/DatasetServiceImplTest.java]: Use `ArgumentMatchers.eq` to make assertions stronger

File: src/main/java/org/ibp/api/rest/dataset/DatasetResource.java [hunk 2]
```diff
@@ -207,6 +212,23 @@ public long getFilteredCount() {
 		return new ResponseEntity<>(observationUnitTable, HttpStatus.OK);
 	}
 
+	@ApiOperation(value = "It will retrieve all the observation units in a simple JSON array table format, this will be used as input data for Visualization tool.",
+		notes = "It will retrieve all data of factors and variate at observation/sub-observation level. Returns data as a simple JSON Array table format.")
+	@RequestMapping(value = "/{cropname}/studies/{studyId}/datasets/{datasetId}/observationUnits/visualization", method = RequestMethod.POST)
+	@ResponseBody
+	public ResponseEntity<List<Map<String, Object>>> getObservationUnitTableAsJSONArray(@PathVariable final String cropname, //
```
Comment (darla-leafnode) [inline src/main/java/org/ibp/api/rest/dataset/DatasetResource.java]: Omit in comment:
> this will be used as input data for Visualization tool.

Comment (darla-leafnode) [reply src/main/java/org/ibp/api/rest/dataset/DatasetResource.java]: It's important to include in API notes that only variables (let's not use factors and variates anymore) specified in `filterColumns` will be retrieved in the map

Comment (darla-leafnode) [reply src/main/java/org/ibp/api/rest/dataset/DatasetResource.java]: **visualization** as part of the endpoint is unclear/improper. What does it mean? The service should be agnostic of where the data is being/or will be used. 

We can rename:
1. method to **getObservationUnitsAsMapList**
2. endpoint to /{cropname}/studies/{studyId}/datasets/{datasetId}/observationUnits/mapList

File: src/main/java/org/ibp/api/java/impl/middleware/dataset/DatasetServiceImpl.java [hunk 3]
```diff
@@ -328,6 +330,19 @@ void validateStudyDatasetAndInstances(
 		return list;
 	}
 
+	@Override
+	public List<Map<String, Object>> getObservationUnitRowsAsListMap(
```
Comment (darla-leafnode) [inline src/main/java/org/ibp/api/java/impl/middleware/dataset/DatasetServiceImpl.java]: More correct: **getObservationUnitRowsAsMapList**

**Comment by abatac** [comment]
Hi @darla-leafnode, I'm done addressing all the comments. Please check.

**Pull request merged**